graphql/http.go

```diff
@@ -0,0 +1,97 @@
+package graphql
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"github.com/samsarahq/thunder/batch"
+	"github.com/samsarahq/thunder/reactive"
+)
+
+func CreateHTTPHandler(schema *Schema, middlewares ...MiddlewareFunc) http.Handler {
+	return &handler{
+		schema:      schema,
+		middlewares: middlewares,
+	}
+}
+
+type handler struct {
+	schema      *Schema
+	middlewares []MiddlewareFunc
+}
+
+type httpPostBody struct {
+	Query     string                 `json:"query"`
+	Variables map[string]interface{} `json:"variables"`
+}
+
+type httpResponse struct {
+	Data   interface{} `json:"data"`
+	Errors []string    `json:"errors"`
+}
+
+func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	writeResponse := func(value interface{}, err error) {
+		response := httpResponse{Errors: []string{err.Error()}}
+		responseJSON, err := json.Marshal(response)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		http.Error(w, string(responseJSON), http.StatusOK)
+	}
+
+	var params httpPostBody
+	if err := json.NewDecoder(r.Body).Decode(&params); err != nil {
+		writeResponse(nil, err)
+		return
+	}
+
+	query, err := Parse(params.Query, params.Variables)
+	if err != nil {
+		writeResponse(nil, err)
+		return
+	}
+
+	if err := PrepareQuery(h.schema.Query, query.SelectionSet); err != nil {
+		writeResponse(nil, err)
+		return
+	}
+
+	e := Executor{}
+	runner := reactive.NewRerunner(r.Context(), func(ctx context.Context) (interface{}, error) {
+		ctx = batch.WithBatching(ctx)
+
+		var middlewares []MiddlewareFunc
+		middlewares = append(middlewares, h.middlewares...)
+		middlewares = append(middlewares, func(input *ComputationInput, next MiddlewareNextFunc) *ComputationOutput {
+			output := next(input)
+			output.Current, output.Error = e.Execute(input.Ctx, h.schema.Query, nil, input.ParsedQuery)
+			return output
+		})
+
+		output := runMiddlewares(middlewares, &ComputationInput{
+			Ctx:         ctx,
+			ParsedQuery: query,
+			Query:       params.Query,
+			Variables:   params.Variables,
+		})
+		current, err := output.Current, output.Error
+
+		if err != nil {
+			if extractPathError(err) == context.Canceled {
+				return nil, err
+			}
+
+			writeResponse(nil, err)
+			return nil, err
+		}
+
+		writeResponse(current, nil)
+		return nil, nil
+	}, MinRerunInterval)
+
+	runner.Stop()
```

Comment:
i forget if it's safe to call this immediately after creating the runner

Reply:
I added a wait group since the re-runner starts a new goroutine.